src/update_agent/actor.rs

```diff
@@ -73,52 +74,92 @@ impl Message for RefreshTick {
+
+            let modify_state_outcome = match prev_state.clone() {
+                UpdateAgentState::StartState => {
+                    update_agent_info
+                        .tick_initialize(&mut *agent_state_guard)
+                        .await
+                }
+                UpdateAgentState::Initialized => {
+                    update_agent_info
+                        .tick_report_steady(&mut *agent_state_guard)
+                        .await
+                }
+                UpdateAgentState::ReportedSteady => {
+                    update_agent_info
+                        .tick_check_updates(&mut *agent_state_guard)
+                        .await
+                }
+                UpdateAgentState::NoNewUpdate => {
+                    update_agent_info
+                        .tick_check_updates(&mut *agent_state_guard)
+                        .await
+                }
+                UpdateAgentState::UpdateAvailable((release, _)) => {
+                    let update = release.clone();
+                    update_agent_info
+                        .tick_stage_update(&mut *agent_state_guard, update)
+                        .await
+                }
+                UpdateAgentState::UpdateStaged((release, _)) => {
+                    let update = release.clone();
+                    update_agent_info
+                        .tick_finalize_update(&mut *agent_state_guard, update)
+                        .await
+                }
+                UpdateAgentState::UpdateFinalized(release) => {
+                    let update = release.clone();
+                    update_agent_info
+                        .tick_end(&mut *agent_state_guard, update)
+                        .await
+                }
+                UpdateAgentState::EndState => Ok(()),
+            };
+
+            if modify_state_outcome.is_err() {
+                log::error!("failed to update agent state");
             }
-            UpdateAgentState::EndState => self.nop(),
-        };
 
-        let update_machine = state_action.then(move |_r, actor, ctx| {
-            if let Some(pause) = actor.refresh_delay(prev_state) {
+            let pause = Self::refresh_delay(
+                update_agent_info.steady_interval,
+                &prev_state,
+                &*agent_state_guard,
+            );
+
+            pause
+        };
+        let state_action = state_action.into_actor(self);
+        let update_machine = state_action.then(|pause, actor, ctx| {
+            if let Some(pause) = pause {
                 log::trace!(
                     "scheduling next agent refresh in {} seconds",
                     pause.as_secs()
                 );
                 Self::tick_later(ctx, pause);
             } else {
                 let update_timestamp = chrono::Utc::now();
-                actor.state_changed = update_timestamp;
+                actor.info.state_changed = update_timestamp;
```

Comment:
Not exactly related to this PR, but I noticed that the logic around updating this `state_changed` field is highly coupled with the logic around `refresh_delay()` which is in turn dependent on whether the `state` was actually changed. I think this made sense back in simpler times when `tick_now` if and only if change in state; however, there are now cases where a change in state does not necessarily `tick_now`.
Anyways, IIUC, I think the logic around updating this `state_changed` field should be refactored to a more intuitive location.

Reply:
Just wondering, would this `state_changed` field fit more naturally into the state itself?

Reply:
I thought about that, but `UpdateAgentState` is an enum. Would it be a good idea to attach the data to the enum the same way we attach e.g. `Release` to some of the variants?

Reply:
I was more thinking about the opposite: turning `state` into its own struct and making the enum just one of its field, next to the timestamp.
On a second thought, I'm not seeing right now what we are using this `state_changed` information for. Possibly it's unused at the moment, in which case you can either decide to drop it or to clarify its location and semantics (e.g. so that it could be sanely queried maybe via dbus).